digiroad2-viite/src/test/scala/fi/liikennevirasto/viite/ProjectValidatorSpec.scala

```diff
@@ -1640,16 +1640,90 @@ class ProjectValidatorSpec extends FunSuite with Matchers {
+      )
+      roadwayDAO.create(roadway)
+
+      //create linear locations
+      val linearLocations = Seq(
+        LinearLocation(linearLocationId, 1, 1000l, 0.0, 10.0, SideCode.TowardsDigitizing, 10000000000l,
+          (CalibrationPointReference(Some(0l)), CalibrationPointReference(Some(10l))),
+          geometryNew, LinkGeomSource.ComplementaryLinkInterface,
+          roadwayNumber1, Some(startDate), None),
+
+        LinearLocation(linearLocationId + 1, 2, 2000l, 0.0, 10.0, SideCode.TowardsDigitizing, 10000000000l,
+          (CalibrationPointReference(Some(0l)), CalibrationPointReference(Some(10l))),
+          geometryUnchanged1, LinkGeomSource.ComplementaryLinkInterface,
+          roadwayNumber1, Some(startDate), None),
+
+        LinearLocation(linearLocationId + 2, 3, 3000l, 0.0, 10.0, SideCode.TowardsDigitizing, 10000000000l,
+          (CalibrationPointReference(Some(0l)), CalibrationPointReference(Some(10l))),
+          geometryUnchanged2, LinkGeomSource.ComplementaryLinkInterface,
+          roadwayNumber1, Some(startDate), None)
+      )
+
+      linearLocationDAO.create(linearLocations)
+
+      //create project and reserve road parts
+      val project = Project(projectId, ProjectState.Incomplete, "f", "s", DateTime.now(), "", DateTime.now(), DateTime.now(), "", Seq(), Seq(), None, None)
+      projectDAO.create(project)
+      projectReservedPartDAO.reserveRoadPart(projectId, 19999L, 1L, "u")
+
+      // create project links
+      projectLinkDAO.create(
+        Seq(
+          util.projectLink(0L, 10L, Combined, projectId, LinkStatus.New, linearLocationId = linearLocationId, discontinuity = Discontinuity.Continuous, linkId = 1000, roadwayId = roadwayId).copy(geometry = geometryNew, originalStartAddrMValue = 0, originalEndAddrMValue = 0),
+          util.projectLink(10L, 20L, Combined, projectId, LinkStatus.UnChanged, linearLocationId = linearLocationId + 1 ,discontinuity = Discontinuity.Continuous, linkId = 2000, roadwayId = roadwayId).copy(geometry = geometryUnchanged1, originalStartAddrMValue = 0, originalEndAddrMValue = 10),
+          util.projectLink(20L, 30L, Combined, projectId, LinkStatus.UnChanged, linearLocationId = linearLocationId + 2 ,discontinuity = Discontinuity.EndOfRoad, linkId = 3000, roadwayId = roadwayId).copy(geometry = geometryUnchanged2, originalStartAddrMValue = 10, originalEndAddrMValue = 20)
+        )
+      )
+
+      // fetch project links
+      val projectLinks = projectLinkDAO.fetchProjectLinks(projectId)
+
+      mockEmptyRoadAddressServiceCalls()
+
+      // call the function that will be tested
+      val validationErrors = projectValidator.checkForInvalidUnchangedLinks(project, projectLinks)
+      validationErrors.size shouldNot be(0)
+      validationErrors.foreach(e => e.validationError.value should be(projectValidator.ValidationErrorList.ErrorInValidationOfUnchangedLinks.value))
```

Comment:
Hmm, montakohan noita tulee, oletan, että vähintään kaksi (yksi per virheellisesti Unchanged lineaarilokaatio/projektilinkki)?
Vai onko tuo checkForInvalid--- rakennettu niin, että palauttaa vain ensimmäisen tjsp.?

Eli kai mä vähän niinku kysyn voisko ton vaidointivirheiden määrän eksplikoida kans (should be(2) tai should be atLeast(2), jos toi atLeast on käytettävissä noin, tai jotain)?

Reply:
Validointi virheitä tulee yksi mutta sen sisällä on mainittuna noi linkId:t (affectedIds tms) eli ton voisi tehdä niin että validationErrors.size should be (1) 
ja sitten vielä tarkistaa että kaikki virheelliset linkit oikeasti löytyvät siitä validointi virheestä